kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala

```diff
@@ -6,30 +6,32 @@
 
 package com.mwam.kafkakewl.domain.validation
 
-import com.mwam.kafkakewl.domain.{Deployments, Namespace, Topology, TopologyDeployments, TopologyId}
+import com.mwam.kafkakewl.domain.*
+import com.mwam.kafkakewl.utils.CollectionExtensions.*
 
+import scala.annotation.unused
 import scala.util.matching.Regex
 
 object TopologyValidation {
   private val namespaceRegex: Regex = """^[a-zA-Z0-9\-_]+(\.([a-zA-Z0-9\-_]+))*$""".r
 
-  def validate(topologyDeploymentsBefore: TopologyDeployments)(deployments: Deployments): ValidationFailures = {
-    deployments.deploy.map(validate).combine
-    // TODO validate topologies and removed topologies together with topologyDeploymentsBefore (via relationships)
+  def validateTopology(topology: Topology): ValidationFailures = {
+    validateTopologyId(topology.id) + validateNamespace(topology.namespace)
   }
 
-  def validate(topology: Topology): ValidationFailures = {
-    validateTopologyId(topology.id) + validateNamespace(topology.namespace)
+  def validateTopologyDependencies(@unused topology: Topology, @unused topologies: Topologies): ValidationFailures = {
+    // TODO implement
+    valid
   }
 
-  def validateNamespace(namespace: Namespace): ValidationFailures = {
-    validationErrorIf(s"namespace ${namespace.quote} must not start or end with '.', '_', '_' and can contain only alphanumeric characters and '.', '-', '_'") {
+  private def validateNamespace(namespace: Namespace): ValidationFailures = {
+    validationErrorIf(s"namespace ${namespace.quote} must not start or end with '.' and can contain only alphanumeric characters and '.', '-', '_'") {
```

Comment:
not sure if I understand this one, we are saying that make sure namespace is not empty and it has nothing after `.` (that's what I got from regex above). But in error message we make it sound like `.` is allowed in namespace

Reply:
looking at tests, the error message seems correct. Maybe I misunderstood what that regex is doing

Reply:
It expects non-empty namspaces to start with a fragment of letters/numbers/hyphen/underscore, then zero more fragments that start with dot and followed by one ore more letters/numbers/hyphen/underscore.

I think it does the job?